### PR TITLE
[Feature] Read binlog to generate change events

### PR DESCRIPTION
**Pull request by banmoy**
## What type of PR is this：
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

The feature to support the storage of binlog has several PRs
* Support to write/read binlog file
* Generate binlog when the ingestion is published
* Cleanup the binlog when expired or oversize
* Load binlog when loading tablet
* Read binlog to generate change events

This is the third PR. Read binlog to generate change events.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by banmoy** [comment]
run starrocks_admit_test

**Review by decster — DISMISSED** [review]

**Comment by banmoy** [comment]
run starrocks_admit_test

File: be/src/storage/binlog_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+#include "column/datum.h"
+#include "storage/chunk_helper.h"
+#include "storage/rowset/rowid_range_option.h"
+#include "storage/rowset/segment_options.h"
+#include "storage/tablet.h"
+
+namespace starrocks {
+
+std::unordered_set<std::string_view> BINLOG_META_SET{BINLOG_OP, BINLOG_VERSION, BINLOG_SEQ_ID, BINLOG_TIMESTAMP};
```
Comment (mofeiatwork) [inline be/src/storage/binlog_reader.cpp:27]: should the global variable be named as `gXxxYyy` ?

Comment (mofeiatwork) [reply be/src/storage/binlog_reader.cpp:27]: 1. is it safe to use a `string_view` ? 
2. why not maintain this state at BinlogManager ?

File: be/src/storage/binlog_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            _rowset = _tablet->get_inc_rowset_by_version(Version(rowset_id, rowset_id));
+        }
+        _rowset->acquire();
+        RETURN_IF_ERROR(_rowset->load());
+        VLOG(3) << "Load rowset for reader: " << _reader_id << ", tablet: " << _tablet->full_name()
+                << ", rowset: " << _rowset->rowset_id() << ", version: " << _rowset->version();
+    }
+
+    int segment_index = _log_entry_info->file_id->segment_index();
+    SegmentSharedPtr seg_ptr = _rowset->segments()[segment_index];
```
Comment (mofeiatwork) [inline be/src/storage/binlog_reader.cpp:257]: validate the index ?

File: be/src/storage/binlog_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    if (_segment_iterator != nullptr) {
+        _segment_iterator->close();
+        _segment_iterator.reset();
+    }
+
+    if (_rowset != nullptr) {
+        _rowset->release();
+        _rowset.reset();
+    }
+
```
Comment (LiShuMing) [inline be/src/storage/binlog_reader.cpp:301]: just reuse `_release_segment_iterator (true)`?

File: be/src/storage/binlog_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    if (log_entry_info->start_seq_id < _next_seq_id) {
+        // for duplicate key, LogEntryInfo#start_row_id is 0
+        rowid_t start_row_id = _next_seq_id - log_entry_info->start_seq_id;
+        SparseRange range(start_row_id, seg_ptr->num_rows());
+        seg_options.rowid_range_option = std::make_shared<RowidRangeOption>(_rowset->rowset_id(), segment_index, range);
+    }
+    auto res = seg_ptr->new_iterator(_data_schema, seg_options);
+    if (!res.ok()) {
+        return res.status();
+    }
```
Comment (LiShuMing) [inline be/src/storage/binlog_reader.cpp:272]: ASSIGN_OR_RETURN?

File: be/src/storage/binlog_reader.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+struct BinlogReaderParams {
+    // Chunk size to read from a segment at a time
+    int chunk_size;
+    // The schema of output from the reader
+    Schema output_schema;
+};
+
+// Column names for metas
+const std::string BINLOG_OP = "_binlog_op";
```
Comment (LiShuMing) [inline be/src/storage/binlog_reader.h:36]: contexpr const?

File: be/src/storage/binlog_manager.h [hunk 6]
```diff
@@ -139,6 +140,18 @@ class BinlogManager : public std::enable_shared_from_this<BinlogManager> {
     // Commit the result of pre-commit, and it's visible for reading
     void commit_ingestion(int64_t version);
 
+    // Register the reader, and return a unique id allocated for this reader.
+    StatusOr<int64_t> register_reader(std::shared_ptr<BinlogReader> reader);
+
+    // Unregister the reader with the given id.
+    void unregister_reader(int64_t reader_id);
```
Comment (LiShuMing) [inline be/src/storage/binlog_manager.h:147]: what's the usage of register/unregister?

File: be/src/storage/binlog_reader.cpp (hunk 2 above)
Comment (banmoy) [reply be/src/storage/binlog_reader.cpp:257]: fixed

File: be/src/storage/binlog_reader.cpp (hunk 3 above)
Comment (banmoy) [reply be/src/storage/binlog_reader.cpp:301]: fixed

File: be/src/storage/binlog_reader.cpp (hunk 4 above)
Comment (banmoy) [reply be/src/storage/binlog_reader.cpp:272]: fixed

**Comment by banmoy** [comment]
run starrocks_admit_test

File: be/src/storage/binlog_reader.cpp (hunk 1 above)
Comment (banmoy) [reply be/src/storage/binlog_reader.cpp]: This variable is not very necessary, and just remove it.

File: be/src/storage/binlog_manager.h (hunk 6 above)
Comment (banmoy) [reply be/src/storage/binlog_manager.h:147]: notify these registered reader when removing expired files, and this logic will be in the next PR

**Review by murphyatwork — DISMISSED** [review]

**Comment by banmoy** [comment]
run starrocks_admit_test

**Review by decster — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**